### PR TITLE
Integrate Document Overlay with the SDK.

### PR DESCRIPTION
**Pull request by ehsannas**

### CODE REVIEW

File: packages/firestore/test/util/helpers.ts
Comment (schmidt-sebastian) [inline packages/firestore/test/util/helpers.ts:305]: 1. Does this have to be exported?
2. I don't like the name "helper". Can this be inline with patchMutation (with a default argument)

File: packages/firestore/test/util/helpers.ts
Comment (schmidt-sebastian) [inline packages/firestore/test/util/helpers.ts:322]:
```suggestion
  const mask = updateMask ? updateMask : parsed.fieldMask.fields;
```

File: packages/firestore/test/util/helpers.ts
Comment (schmidt-sebastian) [inline packages/firestore/test/util/helpers.ts:1011]:
```suggestion
/** 
  * Returns all possible permutations of the given array. 
  * For `[a,b]`, this method returns `[[a,b], [b,a]]`.
  */
```

File: packages/firestore/test/util/helpers.ts
Comment (schmidt-sebastian) [inline packages/firestore/test/util/helpers.ts:1037]:
```suggestion
/**
 * Returns all possible combinations of the given array, including an empty
 * array. For `[a,b,c ]` this method returns `[[], [a], [a,b], [a, c], [b, c], [a, b, c]`.
 */
```

**Review by schmidt-sebastian — COMMENTED** [review]
Thanks for doing this. Must have been fun to spot all the code you had to port :/

Note that we cannot merge this into `master` until we have the code that migrates the existing mutations to overlays.

File: packages/firestore/src/local/local_documents_view.ts
Comment (ehsannas) [reply packages/firestore/src/local/local_documents_view.ts]: Done

File: packages/firestore/src/local/local_documents_view.ts
Comment (ehsannas) [reply packages/firestore/src/local/local_documents_view.ts]: Done

File: packages/firestore/src/local/local_documents_view.ts
Comment (ehsannas) [reply packages/firestore/src/local/local_documents_view.ts]: Done

File: packages/firestore/src/model/mutation.ts
Comment (schmidt-sebastian) [reply packages/firestore/src/model/mutation.ts]: Yes

File: packages/firestore/src/local/local_store_impl.ts
Comment (schmidt-sebastian) [inline packages/firestore/src/local/local_store_impl.ts:327]:
```suggestion
      // TODO(Overlay): Is there a better way to determine this? Using the
      // document version does not work because local mutations set them back to 0.
```

**Review by schmidt-sebastian — COMMENTED** [review]
Thanks for the updated. Please hold off on merging until we have come up with a migration strategy.

File: packages/firestore/src/local/local_documents_view.ts
Comment (ehsannas) [reply packages/firestore/src/local/local_documents_view.ts:186]: Done

File: packages/firestore/src/local/local_documents_view.ts
Comment (ehsannas) [reply packages/firestore/src/local/local_documents_view.ts:230]: Done

File: packages/firestore/src/local/local_documents_view.ts
Comment (ehsannas) [reply packages/firestore/src/local/local_documents_view.ts:233]: Done

File: packages/firestore/src/model/mutation.ts
Comment (ehsannas) [reply packages/firestore/src/model/mutation.ts:583]: Done

**Comment by ehsannas** [comment]
I'll create a new PR against a new `overlay` branch once this PR is approved.

File: packages/firestore/test/unit/local/query_engine.test.ts
Comment (schmidt-sebastian) [reply packages/firestore/test/unit/local/query_engine.test.ts]: I am not sure why it was written this way. I think the test coverage here is equivalent and re-using the existing test pattern removes quite a bit of fragile code (and it also avoids the same question in the C++ port).

File: packages/firestore/src/local/local_documents_view.ts
Comment (schmidt-sebastian) [inline packages/firestore/src/local/local_documents_view.ts:233]: FWIW I know this is my suggested code but today's Sebastian would suggest `masks.get(key) || FieldMask.empty()`. The end behavior in this case is the same, but the transpiled JS is much shorter.

`a || b` translates to just `a || b`, but unfortunately also returns 'b' for any value that evaluates to `false` (0, "", null, undefined). `a ?? b` is the better version of this as it does not return `b` for 0 and "", but it translates to `a !== null && a !== void 0 ? a : b`

See https://www.typescriptlang.org/play?#code/DYUwLgBAhg3AUKSAjeiIGMIF5oQPx4QoLgQAm2uAPlUUA

(you could certainly argue that the impact on code size here is negligible).

File: packages/firestore/src/local/local_documents_view.ts
Comment (schmidt-sebastian) [inline packages/firestore/src/local/local_documents_view.ts:237]: Same as above.

**Review by schmidt-sebastian — APPROVED** [review]

File: packages/firestore/src/local/local_documents_view.ts
Comment (ehsannas) [reply packages/firestore/src/local/local_documents_view.ts]: Done

File: packages/firestore/src/local/local_documents_view.ts
Comment (ehsannas) [reply packages/firestore/src/local/local_documents_view.ts]: Done

**Comment by ehsannas** [comment]
I'll close this as we don't want to merge it into master yet. I'd rather land the squashed changes in to a new branch (`firestore-overlays`) done via #6123

**Pull request closed without merging**
